exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/conversion/DrillCalciteCatalogReader.java

```diff
@@ -0,0 +1,213 @@
+   * @return table instance, null otherwise
+   * @throws UserException if temporary tables usage is disallowed
+   */
+  @Override
+  public Prepare.PreparingTable getTable(List<String> names) {
+    checkIfWeCanProceedForTemporaryTable(names);
+    Prepare.PreparingTable table = super.getTable(names);
+    DrillTable drillTable;
+    if (table != null && (drillTable = table.unwrap(DrillTable.class)) != null) {
+      drillTable.setOptions(session.getOptions());
+      drillTable.setTableMetadataProviderManager(tableCache.getUnchecked(DrillTableKey.of(names, drillTable)));
+    }
+    return table;
+  }
+
+  private void checkIfWeCanProceedForTemporaryTable(List<String> names) {
+    String originalTableName;
+    if (!allowTemporaryTables &&
+        (originalTableName = session.getOriginalTableNameFromTemporaryTable(names.get(names.size() - 1))) != null) {
+      throw UserException
+          .validationError()
+          .message("Temporary tables usage is disallowed. Used temporary table name: [%s].", originalTableName)
+          .build(SqlConverter.logger);
+    }
+  }
+
+  @Override
+  public List<List<String>> getSchemaPaths() {
+    if (useRootSchema.getAsBoolean()) {
+      return ImmutableList.of(ImmutableList.of());
+    }
+    return super.getSchemaPaths();
+  }
+
+  /**
+   * Checks if the schema provided is a valid schema:
+   * <li>schema is not indicated (only one element in the names list)<li/>
+   *
+   * @param names list of schema and table names, table name is always the last element
+   * @throws UserException if the schema is not valid.
+   */
+  void isValidSchema(List<String> names) throws UserException {
+    List<String> schemaPath = Util.skipLast(names);
+
+    for (List<String> currentSchema : getSchemaPaths()) {
+      List<String> fullSchemaPath = new ArrayList<>(currentSchema);
+      fullSchemaPath.addAll(schemaPath);
+      CalciteSchema schema = SqlValidatorUtil.getSchema(getRootSchema(),
+          fullSchemaPath, nameMatcher());
+
+      if (schema != null) {
+       return;
+      }
+    }
+    SchemaUtilites.throwSchemaNotFoundException(defaultSchemaSupplier.get(), schemaPath);
+  }
+
+  /**
+   * We should check if passed table is temporary or not if:
+   * <li>schema is not indicated (only one element in the names list)<li/>
+   * <li>current schema or indicated schema is default temporary workspace<li/>
+   *
+   * Examples (where dfs.tmp is default temporary workspace):
+   * <li>select * from t<li/>
+   * <li>select * from dfs.tmp.t<li/>
+   * <li>use dfs; select * from tmp.t<li/>
+   *
+   * @param names             list of schema and table names, table name is always the last element
+   * @param defaultSchemaPath current schema path set using USE command
+   * @param drillConfig       drill config
+   * @return true if check for temporary table should be done, false otherwise
+   */
+  private boolean mightBeTemporaryTable(List<String> names, String defaultSchemaPath, DrillConfig drillConfig) {
```

Comment:
`needsTempTableCheck` as name? The present name suggests that this does the check.